File: openc3/lib/openc3/interfaces/http_client_interface.rb
# encoding: ascii-8bit

# Copyright 2024 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/interfaces/interface'
require 'openc3/config/config_parser'
require 'faraday'
require 'faraday/follow_redirects'
require 'openc3/accessors/http_accessor'

# TODO: Header Log Filtering? File Uploads? Authorization from secrets?

module OpenC3
  class HttpClientInterface < Interface
    # @param hostname [String] HTTP/HTTPS server to connect to
    # @param port [Integer] HTTP/HTTPS port
    # @param protocol [String] http or https
    def initialize(hostname, port = 80, protocol = 'http', write_timeout = 5, read_timeout = nil,
                   connect_timeout = 5, include_request_in_response = false)
      super()
      @hostname = hostname
      @port = Integer(port)
      @protocol = protocol
      if (@port == 80 and @protocol == 'http') or (@port == 443 and @protocol == 'https')
        @url = "#{@protocol}://#{@hostname}"
      else
        @url = "#{@protocol}://#{@hostname}:#{@port}"
      end
      @write_timeout = ConfigParser.handle_nil(write_timeout)
      @write_timeout = Float(@write_timeout) if @write_timeout
      @read_timeout = ConfigParser.handle_nil(read_timeout)
      @read_timeout = Float(@read_timeout) if @read_timeout
      @connect_timeout = ConfigParser.handle_nil(connect_timeout)
      @connect_timeout = Float(@connect_timeout) if @connect_timeout
      @include_request_in_response = ConfigParser.handle_true_false(include_request_in_response)
      @response_queue = Queue.new
    end

    def connection_string
      return @url
    end

    # Connects the interface to its target(s)
    def connect
      # Per https://github.com/lostisland/faraday/blob/main/lib/faraday/options/env.rb
      # :timeout       - time limit for the entire request (Integer in seconds)
      # :open_timeout  - time limit for just the connection phase (e.g. handshake) (Integer in seconds)
      # :read_timeout  - time limit for the first response byte received from the server (Integer in seconds)
      # :write_timeout - time limit for the client to send the request to the server (Integer in seconds)
      request = {}
      request['open_timeout'] = @connect_timeout if @connect_timeout
      request['read_timeout'] = @read_timeout if @read_timeout
      request['write_timeout'] = @write_timeout if @write_timeout
      @http = Faraday.new(request: request) do |f|
        f.response :follow_redirects # use Faraday::FollowRedirects::Middleware
        f.adapter :net_http # adds the adapter to the connection, defaults to `Faraday.default_adapter`
      end
      super()
    end

    # Whether the interface is connected to its target(s)
    # But in this case it is basically whether connect was called
    def connected?
      if @http
        return true
      else
        return false
      end
    end

    # Disconnects the interface from its target(s)
    def disconnect
      @http.close if @http
      @http = nil
      # Clear the response queue
      while @response_queue.length > 0
        @response_queue.pop
      end
      super()
      # Push nil to cause read_interface to return nil
      @response_queue.push(nil)
    end

    # Called to convert a packet into a data buffer. Write protocols then
    # potentially modify the data in their write_data methods. Finally
    # write_interface is called to send the data to the target.
    #
    # @param packet [Packet] Packet to extract data from
    # @return data, extra
    def convert_packet_to_data(packet)
      extra = packet.extra
      extra ||= {}
      data = packet.buffer(true) # Copy buffer so logged command isn't modified
      extra['HTTP_URI'] = URI("#{@url}#{packet.read('HTTP_PATH')}").to_s
      # Store the target name for use in identifying the response
      extra['HTTP_REQUEST_TARGET_NAME'] = packet.target_name
      return data, extra
    end

    # Calls the appropriate HTTP method using Faraday
    def write_interface(data, extra = nil)
      extra ||= {}
      queries = extra['HTTP_QUERIES']
      queries ||= {}
      headers = extra['HTTP_HEADERS']
      headers ||= {}
      uri = extra['HTTP_URI']
      method = extra['HTTP_METHOD']

      resp = nil
      case method
      when 'get'
        resp = @http.get(uri) do |req|
          req.params = queries
          req.headers = headers
        end
      when 'put'
        resp = @http.put(uri) do |req|
          req.params = queries
          req.headers = headers
          req.body = data
        end
      when 'delete'
        resp = @http.delete(uri) do |req|
          req.params = queries
          req.headers = headers
        end
      when 'post'
        resp = @http.post(uri) do |req|
          req.params = queries
          req.headers = headers
          req.body = data
        end
      else
        raise "Unsupported HTTP Method: #{method}"
      end

      # Normalize Response into simple hash
      response_data = nil
      response_extra = {}
      if resp
        # We store the request data and extra under HTTP_REQUEST
        # This can optionally be returned in the response
        # but is also used to identify the response target
        response_extra['HTTP_REQUEST'] = [data, extra]
        if resp.headers and resp.headers.length > 0
          response_extra['HTTP_HEADERS'] = resp.headers
        end
        response_extra['HTTP_STATUS'] = resp.status
        response_data = resp.body
        response_data ||= '' # Ensure an empty string if we got a response but no data
      end

      @response_queue.push([response_data, response_extra])

      write_interface_base(data, extra)
      return data, extra
    end

    # Returns the response data and extra from the interface
    # which was queued up by the write_interface method.
    # Read protocols can then potentially modify the data in their read_data methods.
    # Then convert_data_to_packet is called to convert the data into a Packet object.
    # Finally the read protocols read_packet methods are called.
    def read_interface
      # This blocks until a response is available
      data, extra = @response_queue.pop
      return nil if data.nil?

      read_interface_base(data, extra)
      return data, extra
    end

    # Called to convert the read data into a OpenC3 Packet object
    #
    # @param data [String] Raw packet data
    # @param extra [Hash] Contains the following keys:
    #   HTTP_HEADERS - Hash of response headers
    #   HTTP_STATUS - Integer response status code
    #   HTTP_REQUEST - [data, extra]
    #     where data is the request data and extra contains:
    #       HTTP_REQUEST_TARGET_NAME - String request target name
    #       HTTP_URI - String request URI based on HTTP_PATH
    #       HTTP_PATH - String request path
    #       HTTP_METHOD - String request method
    #       HTTP_PACKET - String response packet name
    #       HTTP_ERROR_PACKET - Optional string error packet name
    #       HTTP_QUERIES - Optional hash of request queries
    #       HTTP_HEADERS - Optional hash of request headers
    # @return [Packet] OpenC3 Packet with buffer filled with data
    def convert_data_to_packet(data, extra = nil)
      packet = Packet.new(nil, nil, :BIG_ENDIAN, nil, data.to_s)
      packet.accessor = HttpAccessor.new(packet)
      # Grab the request extra set in the write_interface method
      request_extra = extra['HTTP_REQUEST'][1] if extra and extra['HTTP_REQUEST']
      if request_extra
        # Identify the response target
        request_target_name = request_extra['HTTP_REQUEST_TARGET_NAME']
        if request_target_name
          request_target_name = request_target_name.to_s.upcase
          response_packet_name = request_extra['HTTP_PACKET']
          error_packet_name = request_extra['HTTP_ERROR_PACKET']
          # HTTP_STATUS was set in the base extra
          status = extra['HTTP_STATUS'].to_i
          if status >= 300 and error_packet_name
            # Handle error special case response packet
            packet.target_name = request_target_name
            packet.packet_name = error_packet_name.to_s.upcase
          else
            if response_packet_name
              packet.target_name = request_target_name
              packet.packet_name = response_packet_name.to_s.upcase
            end
          end
        end
        unless @include_request_in_response
          extra.delete("HTTP_REQUEST")
        end
      end
      packet.extra = extra
      return packet
    end

    def details
      result = super()
      result['url'] = @url
      result['write_timeout'] = @write_timeout
      result['read_timeout'] = @read_timeout
      result['connect_timeout'] = @connect_timeout
      result['include_request_in_response'] = @include_request_in_response
      result['request_queue_length'] = @request_queue.length
      return result
    end
  end
end

File: openc3/python/openc3/interfaces/http_client_interface.py
# Copyright 2024 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import queue
import requests
from openc3.config.config_parser import ConfigParser
from openc3.interfaces.interface import Interface
from openc3.accessors.http_accessor import HttpAccessor
from openc3.packets.packet import Packet


class HttpClientInterface(Interface):
    """
    HttpClientInterface is a class that provides an interface for making HTTP
    requests using the requests library.
    """

    def __init__(
        self,
        hostname,
        port=80,
        protocol="http",
        write_timeout=None,
        read_timeout=None,
        connect_timeout=5,
        include_request_in_response=False,
    ):
        """
        Initializes the HTTPClientInterface with the given parameters.
        Args:
            hostname (str): The hostname of the server.
            port (int, optional): The port number to connect to. Defaults to 80.
            protocol (str, optional): The protocol to use ('http' or 'https'). Defaults to "http".
            write_timeout (None, optional): Present to match Ruby parameters but not used.
            read_timeout (float or None, optional): The timeout for reading operations in seconds. Defaults to None.
            connect_timeout (float or None, optional): The timeout for connection operations in seconds. Defaults to 5.
            include_request_in_response (bool, optional): Whether to include the request in the response. Defaults to False.
        """
        super().__init__()
        self.hostname = hostname
        self.port = int(port)
        self.protocol = protocol
        if (self.port == 80 and self.protocol == "http") or (self.port == 443 and self.protocol == "https"):
            self.url = f"{self.protocol}://{self.hostname}"
        else:
            self.url = f"{self.protocol}://{self.hostname}:{self.port}"
        self.read_timeout = ConfigParser.handle_none(read_timeout)
        if self.read_timeout:
            self.read_timeout = float(self.read_timeout)
        self.connect_timeout = ConfigParser.handle_none(connect_timeout)
        if self.connect_timeout:
            self.connect_timeout = float(self.connect_timeout)
        self.include_request_in_response = ConfigParser.handle_true_false(include_request_in_response)

        self.response_queue = queue.Queue()

    def connection_string(self):
        """Returns the url."""
        return self.url

    def connect(self):
        """
        Initializes an HTTP session and then calls the parent class's connect method.
        """
        self.http = requests.Session()
        super().connect()

    def connected(self):
        """
        Check if the HTTP client is connected.

        Returns:
            bool: True if the HTTP client is connected, False otherwise.
        """
        if self.http:
            return True
        else:
            return False

    # Disconnects the interface from its target(s)
    def disconnect(self):
        """
        Disconnects the HTTP client interface.

        This method closes the HTTP connection if it exists, sets the HTTP client to None,
        clears the response queue, calls the superclass's disconnect method, and unblocks
        the response queue to allow the read_interface method to return.
        """
        if self.http:
            self.http.close
        self.http = None
        while not self.response_queue.empty():
            self.response_queue.get_nowait()
        super().disconnect()
        self.response_queue.put((None, None))

    def convert_packet_to_data(self, packet):
        """
        Converts a packet to data and extracts additional information.

        Args:
            packet (Packet): The packet to be converted.
        Returns:
            tuple: A tuple containing:
                - data (bytes): The buffer data from the packet.
                - extra (dict): A dictionary containing additional information extracted from the packet.
                    - "HTTP_URI" (str): The full HTTP URI constructed from the base URL and the packet's HTTP path.
                    - "HTTP_REQUEST_TARGET_NAME" (str): The target name of the HTTP request.
        """

        extra = packet.extra
        extra = extra or {}
        data = packet.buffer  # Copy buffer so logged command isn't modified
        extra["HTTP_URI"] = f"{self.url}{packet.read('HTTP_PATH')}"
        extra["HTTP_REQUEST_TARGET_NAME"] = packet.target_name
        return data, extra

    # Calls the http request method to send the data to the target
    def write_interface(self, data, extra=None):
        """
        Sends the data to the target using an HTTP request.

        Args:
            data (bytes): The data to be sent.
            extra (dict, optional): Additional parameters for the HTTP request. Defaults to None.
                - HTTP_QUERIES: Query parameters for the HTTP request.
                - HTTP_HEADERS: Headers for the HTTP request.
                - HTTP_URI: The URI for the HTTP request.
                - HTTP_METHOD: The HTTP method (e.g., GET, POST).

        Returns:
            tuple: The data and extra parameters.
        """
        extra = extra or {}
        params = extra.get("HTTP_QUERIES")
        headers = extra.get("HTTP_HEADERS")
        uri = extra["HTTP_URI"]
        method = extra["HTTP_METHOD"]

        resp = self.http.request(
            method,
            uri,
            params=params,
            headers=headers,
            data=data,
            stream=False,
            timeout=(self.connect_timeout, self.read_timeout),
        )

        # Normalize Response into simple hash
        response_data = None
        response_extra = {}
        response_extra["HTTP_REQUEST"] = [data, extra]
        if resp.headers and len(resp.headers) > 0:
            # Cast headers to a dictionary so it can be serialized
            # because the request library returns CaseInsensitiveDict
            response_extra["HTTP_HEADERS"] = dict(resp.headers)
        response_extra["HTTP_STATUS"] = resp.status_code
        response_data = bytearray(resp.text, encoding="utf-8")

        self.response_queue.put((response_data, response_extra))
        self.write_interface_base(data, extra)
        return data, extra

    def read_interface(self):
        """
        Returns the response data and extra parameters from the interface,
        which were queued up by the write_interface method.
        Read protocols can then potentially modify the data in their read_data methods.
        Then convert_data_to_packet is called to convert the data into a Packet object.
        Finally the read protocols read_packet methods are called.

        Returns:
            tuple: The response data and extra parameters.
        """
        data, extra = self.response_queue.get(block=True)
        if data is None:
            return data, extra
        self.read_interface_base(data, extra)
        return data, extra

    def convert_data_to_packet(self, data, extra=None):
        """
        Converts the read data into an OpenC3 Packet object.

        Args:
            data (str): Raw packet data.
            extra (dict, optional): Additional parameters for the packet. Defaults to None.
                - HTTP_HEADERS: Hash of response headers.
                - HTTP_STATUS: Integer response status code.
                - HTTP_REQUEST: [data, extra] where data is the request data and extra contains:
                    - HTTP_REQUEST_TARGET_NAME: String request target name.
                    - HTTP_URI: String request URI based on HTTP_PATH.
                    - HTTP_PATH: String request path.
                    - HTTP_METHOD: String request method.
                    - HTTP_PACKET: String response packet name.
                    - HTTP_ERROR_PACKET: Optional string error packet name.
                    - HTTP_QUERIES: Optional hash of request queries.
                    - HTTP_HEADERS: Optional hash of request headers.

        Returns:
            Packet: OpenC3 Packet with buffer filled with data.
        """
        packet = Packet(None, None, "BIG_ENDIAN", None, data)
        packet.accessor = HttpAccessor(packet)
        # Grab the request extra set in the write_interface method
        request_extra = None
        if extra and extra["HTTP_REQUEST"]:
            request_extra = extra["HTTP_REQUEST"][1]
        if request_extra is not None:
            # Identify the response
            request_target_name = request_extra.get("HTTP_REQUEST_TARGET_NAME")
            if request_target_name is not None:
                request_target_name = str(request_target_name).upper()
                response_packet_name = request_extra.get("HTTP_PACKET")
                error_packet_name = request_extra.get("HTTP_ERROR_PACKET")
                # HTTP_STATUS was set in the base extra
                status = int(extra["HTTP_STATUS"])
                if status >= 300 and error_packet_name is not None:
                    # Handle error special case response packet
                    packet.target_name = request_target_name
                    packet.packet_name = str(error_packet_name).upper()
                elif response_packet_name is not None:
                    packet.target_name = request_target_name
                    packet.packet_name = str(response_packet_name).upper()

            if not self.include_request_in_response:
                extra.pop("HTTP_REQUEST", None)
        packet.extra = extra
        return packet
